### PR TITLE
Add necessary options to run salt-ssh as a normal user

### PR DESCRIPTION
**Pull request by bugfood**
### What does this PR do?

This adds the `--cachedir` and `--pki-dir` options necessary for running salt-ssh as a normal user without a config file.

The commits are by @crhan; I had to only make a minor change to fix a merge conflict.

The original PR for this resulted in a merge to the `develop` branch:
https://github.com/saltstack/salt/pull/50802

The changes never got into `master`, though.

### What issues does this PR fix or reference?
Fixes:
https://github.com/saltstack/salt/issues/44073

### Previous Behavior
Errors from salt-ssh when run as a normal user.

### New Behavior
Errors are still present by default, but the user has the option to fix the problem by specifying appropriate `--cachedir` and `--pki-dir` parameters.

Please note that this code was already accepted in:
https://github.com/saltstack/salt/pull/50802

@cachedout, since you approved and merged the original PR, can you please do the same for this one?

### CODE REVIEW

**Review by Ch3LL — CHANGES_REQUESTED** [review]
This will require a changelog file.

Also can we add some test coverage for these new arguments for salt-ssh. A good place to start would be in this test directory: tests/pytests/integration/ssh/

**Comment by bugfood** [comment]
> This will require a changelog file.
> 
> Also can we add some test coverage for these new arguments for salt-ssh. A good place to start would be in this test directory: tests/pytests/integration/ssh/

@Ch3LL I'm quite unfamiliar with the salt test suite and I'll need more help with that. Here's what I've found out so far:

I can run an individual integration test file by:
```
PATH=/usr/sbin:$PATH nox -e 'test-3(coverage=False)' -- --run-slow -v tests/pytests/integration/ssh/test_grains.py:text_experiment
```

I'm using `test_grains.py` to add experimental code, for now. I can add this experimental test, just to see what happens:
```
def test_experiment(salt_ssh_cli):
    ret = salt_ssh_cli.run(
        "--pki-dir", "/tmp/pki", "--cachedir", "/tmp/cache", "grains.items"
    )
    assert ret.returncode == 0
```
The results are:
1. The test passes.
2. The directory "/tmp/cache" is created and populated, so we know that the option is being respected.
3. The directory "/tmp/pki" is NOT created, so we know that the option is being ignored.

That last problem seems to be specific to the test suite, since the option works ok for me in manual testing.

The test parameter `salt_ssh_cli` is an object: `<class 'saltfactories.cli.ssh.SaltSsh'>` which is stringified as:
```
SaltSsh(id='master-dn0A5c', config_file='/tmp/stsuite/master-dn0A5c/conf/master', config_dir='/tmp/stsuite/master-dn0A5c/conf', python_executable='/usr/local/src/salt/.nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false/bin/python', cwd=PosixPath('/usr/local/src/salt'), slow_stop=True, timeout=180, script_name='/tmp/stsuite/scripts/cli_salt_ssh.py', base_script_args=['--ignore-host-keys'], display_name=None, roster_file=PosixPath('/tmp/stsuite/master-dn0A5c/conf/roster'), client_key='/tmp/stsuite/sshd/client_key', target_host='localhost', ssh_user='bugfood')
```

It appears that the creation and configuration of salt-ssh, including the built-in master, is abstracted away from the integration tests. I don't know how to configure that independently.

**Comment by Ch3LL** [comment]
Can you create a `salt_ssh_cli` pytest fixture in your test file that adds the arguments in `base_script_args=`. Similar to this test: tests/pytests/integration/ssh/test_ssh_setup.py. Does that work? If not i can help dive in.

**Comment by bugfood** [comment]
I tried to revisit this earlier, but I hit a snag: tests that worked for me before started timing out.

After troubleshooting, I found that the test suite was hitting a permissions error for a file in the thin_dir; since I am managing this host via salt-ssh, the thin_dir existed and some files therein had root ownership.

I think the thin_dir should be set differently during the integration tests in order to avoid this conflict. I may make a patch for that another time.

I'll keep working on this as I have time.

**Comment by Ch3LL** [comment]
Thanks, let me know if I can help with anything else to help progress this PR.

**Review by cmcmarrow — CHANGES_REQUESTED** [review]
Like what @Ch3LL said we need a change log and a test. Thank you for the pull request.

**Comment by bugfood** [comment]
@Ch3LL thanks. I've been looking at the tests more and I think I have a better grasp of what is going on via the test fixtures.

I'm still having trouble writing a test, though, and I think understand why my previous attempt failed. You can see my experimental work in:
https://github.com/bugfood/salt/tree/salt-ssh-params-experimental
In particular, the latest commit:
https://github.com/bugfood/salt/commit/salt-ssh-params-experimental

What I have found is that I am able to alter the `pki_dir` in the master configuration by means of modifying the config_overrides in `tests/pytests/conftest.py`. I do that by adding a new fixture `salt_master_extra_config_overrides()`. That isn't really what I want to do, though--all it does is modify the configuration without actually testing the command-line parameter.

I haven't been able to make the command-line parameter be respected within the tests. I think this is because the test fixtures present the test function with a salt master which is already configured and running. The test function can't make proper use of any command-line options that affect the master--by that point, the master is already running, so it's too late.

I don't think I'm approaching this the right way and I'll need some more help with how to proceed.

**Comment by Ch3LL** [comment]
@s0undt3ch do you know off the top of your head a good way to test these new arguments with salt-ssh and the test fixtures?

File: salt/utils/parsers.py [hunk 1]
```diff
@@ -3370,6 +3371,11 @@ def _mixin_setup(self):
                 "into the default roster file (flat)."
             ),
         )
+        auth_group.add_option(
+            "--pki-dir",
+            dest="pki_dir",
+            help=("Set the directory to load the pki keys."),
```
Comment (s0undt3ch) [inline salt/utils/parsers.py:3377]:
```suggestion
            help="Set the directory to load the pki keys.",
```

**Review by s0undt3ch — CHANGES_REQUESTED** [review]
It looks like adding the CLI flag/mixin was not enough for salt-ssh to pick up that config.

The reason it might work by hand is because there's no master config file?

I'd look at the SaltSSH parser class and see if the merged options are getting the new value.
If it gets the new value, it means it's being ignored somewhere else in the salt-ssh execution, over what's defined on the master config file.

File: salt/utils/parsers.py (hunk 1 above)
Comment (bugfood) [reply salt/utils/parsers.py:3377]: Sure, I can make that change when I revise the PR.

**Comment by bugfood** [comment]
> It looks like adding the CLI flag/mixin was not enough for salt-ssh to pick up that config.
> 
> The reason it might work by hand is because there's no master config file?

When I test by hand, I do have a master config file (for now--if I can remove it later, I will). The results are:

* Without `pki_dir` in the config file and without `--pki-dir` on the command line, salt-ssh fails: `salt-ssh could not be run because it could not generate keys.`
* With `pki_dir` in the config file, salt-ssh uses the specified dir.
* With `--pki-dir` on the command line, salt-ssh uses the specified dir.
* With both `pki_dir` in the config file and `--pki-dir` on the command line, and with different values specified for each, salt-ssh uses the dir specified on the command line.

That's the behavior I would expect. Maybe it's working for the wrong reasons--I wouldn't know. It is working, though.

> I'd look at the SaltSSH parser class and see if the merged options are getting the new value. If it gets the new value, it means it's being ignored somewhere else in the salt-ssh execution, over what's defined on the master config file.

Do you mean `class SaltSSHOptionParser` in `salt/utils/parsers.py`? I added some debugging to that file:
```
diff --git a/salt/utils/parsers.py b/salt/utils/parsers.py
index 739b8f2057..1ff077b14c 100644
--- a/salt/utils/parsers.py
+++ b/salt/utils/parsers.py
@@ -225,6 +225,7 @@ class OptionParser(optparse.OptionParser):
             )
 
         salt.utils.process.appendproctitle("MainProcess")
+        print("parse_args: pki_dir:", options.pki_dir)
         # Retain the standard behavior of optparse to return options and args
         return options, args
 
@@ -3429,6 +3430,7 @@ class SaltSSHOptionParser(
     def setup_config(self):
         opts = config.master_config(self.get_config_file_path())
         salt.features.setup_features(opts)
+        print("setup_config: pki_dir:", opts['pki_dir'])
         return opts
 ```
 
When I have `pki_dir` specified in the config and `--pki-dir` specified on the command line, that gives output:
```
setup_config: pki_dir: /home/bugfood/git/admin/salt-ssh/pki-in-config-not-used
parse_args: pki_dir: /home/bugfood/git/admin/salt-ssh/pki
```
(and salt-ssh uses `/home/bugfood/git/admin/salt-ssh/pki`, as expected)

-Corey

**Comment by Ch3LL** [comment]
I'll try diving into this in the next week and help figure out why its not working here.

**Comment by Ch3LL** [comment]
When you say you add pki_dir to the config, what config is that exactly?

When testing this I could only see `pki_dir` work when set on the cli and in the Saltfile. When I tried setting it in the master config it does not work.

**Comment by bugfood** [comment]
> When you say you add pki_dir to the config, what config is that exactly?

For manual testing, I have a small `master` file:
```
cachedir: /home/bugfood/git/admin/salt-ssh/cache
pki_dir: /home/bugfood/git/admin/salt-ssh/pki

file_roots:
    base:
        - /home/bugfood/git/admin/salt
```
...which I reference via `salt-ssh --config-dir=/home/bugfood/git/admin/salt-ssh`
 
> When testing this I could only see `pki_dir` work when set on the cli and in the Saltfile. When I tried setting it in the master config it does not work.

You mean you are testing this PR and you do not see it work when setting `pki_dir` in the `master` config file? Does the same manual test work on upstream `master`? The code in the PR should not have any effect on the config file parsing, as I understand it.

Thanks,
Corey

**Comment by Ch3LL** [comment]
Apologies for the late reply. I recently got back from PTO and I am trying to catch up on my messages. 

I'm going to set this one aside to dive into again this week, but yes I wasn't seeing the master setting apply. I will check to see if I was doing something wrong and rebase on the head of master. thanks

**Comment by bugfood** [comment]
> Apologies for the late reply. I recently got back from PTO and I am trying to catch up on my messages.
> 
> I'm going to set this one aside to dive into again this week, but yes I wasn't seeing the master setting apply. I will check to see if I was doing something wrong and rebase on the head of master. thanks

Ok, thank you. I'm not in a huge hurry, but if you can help with this at some point, then I will try to work on the necessary changes, depending on what you find needs to be done.

**Comment by dwoz** [comment]
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.

**Pull request closed without merging**